src/backend/distributed/commands/multi_copy.c

```diff
@@ -2311,46 +2408,84 @@ CitusSendTupleToPlacements(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest
 	/* connections hash is kept in memory context */
 	MemoryContextSwitchTo(copyDest->memoryContext);
 
-	/* get existing connections to the shard placements, if any */
-	shardConnections = GetShardHashConnections(shardConnectionHash, shardId,
-											   &shardConnectionsFound);
-	if (!shardConnectionsFound)
+	shardState = GetShardState(shardId, copyDest->shardStateHash,
+							   copyDest->connectionStateHash, stopOnFailure, &found);
+	if (!found && !copyDest->multiShardCopy &&
+		hash_get_num_entries(copyDest->shardStateHash) == 2)
 	{
-		/*
-		 * Keep track of multi shard accesses before opening connection
-		 * the second shard.
-		 */
-		if (!copyDest->multiShardCopy && hash_get_num_entries(shardConnectionHash) == 2)
-		{
-			Oid relationId = copyDest->distributedRelationId;
+		Oid relationId = copyDest->distributedRelationId;
 
-			/* mark as multi shard to skip doing the same thing over and over */
-			copyDest->multiShardCopy = true;
+		/* mark as multi shard to skip doing the same thing over and over */
+		copyDest->multiShardCopy = true;
 
-			/* error out of conflicting COPY */
-			CheckConflictingParallelCopyAccesses(relationId);
+		/* error out of conflicting COPY */
+		CheckConflictingParallelCopyAccesses(relationId);
 
+		if (MultiShardConnectionType != SEQUENTIAL_CONNECTION)
+		{
 			/* when we see multiple shard connections, we mark COPY as parallel modify */
 			RecordParallelModifyAccess(relationId);
 		}
+	}
 
-		/* open connections and initiate COPY on shard placements */
-		OpenCopyConnections(copyStatement, shardConnections, stopOnFailure,
-							copyOutState->binary);
+	foreach(placementStateCell, shardState->placementStateList)
+	{
+		CopyPlacementState *placementState = lfirst(placementStateCell);
+		CopyConnectionState *connectionState = placementState->connectionState;
+		bool switchOver = false;
 
-		/* send copy binary headers to shard placements */
-		if (copyOutState->binary)
+		if (placementState != connectionState->activePlacementState)
 		{
-			SendCopyBinaryHeaders(copyOutState, shardId,
-								  shardConnections->connectionList);
+			if (placementState->data->len > COPY_SWITCH_OVER_THRESHOLD ||
+				connectionState->activePlacementState == NULL)
```

Comment:
if we make the change in `InitializeCopyShardState` then we can remove `connectionState->activePlacementState == NULL`

Reply:
> if we make the change in `InitializeCopyShardState` then we can remove `connectionState->activePlacementState == NULL`

The reason for initializing this to `NULL` was that we need to call `StartPlacementStateCopyCommand()` for the first placement state somewhere. I could either do this or check the value of `cachedShardStateFound`. I thought defining `activePlacementState` as "the placement for which the COPY has been started" makes sense, and COPY hasn't started before `StartPlacementStateCopyCommand()` is called, so setting it to NULL until here made more sense to me.